yarn-project/end-to-end/src/e2e_p2p/preferred_gossip_network.test.ts

```diff
@@ -0,0 +1,354 @@
+    const validatorEnrs = await Promise.all(validators.map(p => p.getEncodedEnr()));
+    const validatorPeerIds = await Promise.all(validatorEnrs.map(x => ENR.decodeTxt(x!).peerId()));
+
+    // No-discovery validators should only receive P2P gossip from preferred nodes, as that is all they connect to
+    noDiscoveryValidators.forEach(validator => {
+      monitorP2PTraffic(
+        validator,
+        preferredNodePeerIds.map(x => x.toString()),
+      );
+    });
+    // Preferred nodes should only receive P2P gossip from validators
+    preferredNodes.forEach(node => {
+      monitorP2PTraffic(
+        node,
+        validatorPeerIds.map(x => x.toString()),
+      );
+    });
+
+    // We need to `createNodes` before we setup account, because
+    // those nodes actually form the committee, and so we cannot build
+    // blocks without them (since targetCommitteeSize is set to the number of nodes)
+    await t.setupAccount();
+
+    // Send the required number of transactions to each node
+    t.logger.info('Submitting transactions');
+    for (const node of nodes) {
+      const context = await createPXEServiceAndSubmitTransactions(t.logger, node, NUM_TXS_PER_NODE, t.fundedAccount);
+      contexts.push(context);
+    }
+
+    t.logger.info('Waiting for transactions to be mined');
+    // now ensure that all txs were successfully mined
+    await Promise.all(
+      contexts.flatMap((context, i) =>
+        context.txs.map(async (tx, j) => {
+          t.logger.info(`Waiting for tx ${i}-${j}: ${await tx.getTxHash()} to be mined`);
+          return tx.wait({ timeout: WAIT_FOR_TX_TIMEOUT });
+        }),
+      ),
+    );
+    t.logger.info('All transactions mined');
+
+    // Gather signers from attestations downloaded from L1
+    const blockNumber = await contexts[0].txs[0].getReceipt().then(r => r.blockNumber!);
+    const dataStore = ((nodes[0] as AztecNodeService).getBlockSource() as Archiver).dataStore;
+    const [block] = await dataStore.getPublishedBlocks(blockNumber, blockNumber);
+    const payload = ConsensusPayload.fromBlock(block.block);
+    const attestations = block.attestations
+      .filter(a => !a.signature.isEmpty())
+      .map(a => new BlockAttestation(blockNumber, payload, a.signature));
+    const signers = await Promise.all(attestations.map(att => att.getSender().toString()));
+    t.logger.info(`Attestation signers`, { signers });
+
+    expect(signers.length).toEqual(validators.length);
+
+    // Check that the signers found are part of the proposer nodes to ensure the archiver fetched them right
+    const validatorAddresses = validators.flatMap(node =>
+      ((node as AztecNodeService).getSequencer() as SequencerClient).validatorAddresses?.map(a => a.toString()),
+    );
+    t.logger.info(`Validator addresses`, { addresses: validatorAddresses });
+    for (const signer of signers) {
+      expect(validatorAddresses).toContain(signer);
+    }
```

Comment:
I think we can remove this, it's copy paste from another test, right?

Reply:
Oh, I thought this was good to leave in. This ensures that all validators have attested, which ensures that they have all received every tx and block proposal.